python/dolfinx/io.py

```diff
@@ -62,17 +65,210 @@ def read_mesh(self, ghost_mode=cpp.mesh.GhostMode.shared_facet, name="mesh", xpa
+            element_data = gmsh_model.mesh.getElements(dim, tag=entity)
+            element_types, element_tags, node_tags = element_data
+            assert(len(element_types) == 1)
+            # The MSH type of the cells on the element
+            element_type = element_types[0]
+            num_el = len(element_tags[0])
+            # Determine number of local nodes per element to create
+            # the topology of the elements
+            properties = gmsh_model.mesh.getElementProperties(element_type)
+            name, dim, order, num_nodes, local_coords, _ = properties
+            # 2D array of shape (num_elements,num_nodes_per_element) containing
+            # the topology of the elements on this entity
+            # NOTE: GMSH indexing starts with 1 and not zero.
+            element_topology = node_tags[0].reshape(-1, num_nodes) - 1
+
+            # Gather data for each element type and the
+            # corresponding physical markers
+            if element_type in topologies.keys():
+                topologies[element_type]["topology"] =\
+                    numpy.concatenate((topologies[element_type]["topology"],
+                                       element_topology), axis=0)
+                topologies[element_type]["cell_data"] =\
+                    numpy.concatenate((topologies[element_type]["cell_data"],
+                                       numpy.full(num_el, tag)), axis=0)
+            else:
+                topologies[element_type] = {"topology": element_topology,
+                                            "cell_data": numpy.full(num_el, tag)}
+
+    return topologies
+
+
+def extract_geometry(gmsh_model, model_name=None):
+    """
+    For a given gmsh model, extract the mesh geometry
+    as a numpy (N,3) array where the i-th row
+    corresponds to the i-th node in the mesh
+    """
+    if model_name is not None:
+        gmsh_model.setCurrent(model_name)
+    # Get the unique tag and coordinates for nodes
+    # in mesh
+    indices, points, _ = gmsh_model.mesh.getNodes()
+    points = points.reshape(-1, 3)
+    # GMSH indices starts at 1
+    indices -= 1
+    # Sort nodes in geometry according to the unique index
+    perm_sort = numpy.argsort(indices)
+    assert numpy.all(indices[perm_sort] == numpy.arange(len(indices)))
+    return points[perm_sort]
+
+
+def read_from_msh(filename: str, cell_data=False, facet_data=False, gdim=None):
+    """
+    Reads a mesh from a msh-file and returns the dolfin-x mesh.
+    Input:
+        filename: Name of msh file
+        cell_data: Boolean, True of a mesh tag for cell data should be returned
+                   (Default: False)
+        facet_data: Boolean, True if a mesh tag for facet data should be
+                    returned (Default: False)
+        gdim: Geometrical dimension of problem (Default: 3)
+    """
+    if gdim is None:
+        gdim = 3
+    if MPI.COMM_WORLD.rank == 0:
+        import gmsh
+        gmsh.initialize()
```

Comment:
What would happen if a user had called already `gmsh.initialize()` before calling this function?

Reply:
Fixed in latest commit